app/Http/Controllers/ApplicationTimelineController.php

```diff
@@ -14,8 +16,46 @@ class ApplicationTimelineController extends Controller
      * @param  \App\Models\JobApplication $application Incoming Application object.
      * @return \Illuminate\Http\Response
     */
-    public function show(JobApplication $jobApplication)
+    public function show(JobApplication $jobApplication, string $step = null)
     {
+        $requestedStep = $step;
+        $jobApplicationSteps = $jobApplication->jobApplicationSteps();
+        $stepOrder = [
+            'basic',
+            'experience',
+            'skills-intro' => 'skills',
+            'skills',
+            'fit',
+            'review',
+            'submission'
+        ];
+
+
+        // If the applicant has previously reached the requested step
+        // then return them to the step.
+
+        if (
+            $requestedStep !== null &&
+            array_key_exists($requestedStep, $jobApplicationSteps) &&
+            ($jobApplicationSteps[$requestedStep] === 'complete' || $jobApplicationSteps[$requestedStep] === 'error')
+        ) {
+            return view('applicant/application-timeline-root')
+            ->with([
+                'title' => $jobApplication->job_poster->title, // TODO: Check with design what the title should be.
+                'disable_clone_js' => true,
+            ]);
+        }
+
+        // If the applicant has not reached the requested step
+        // in the application then redirect back to last touched step.
+        $reversedStepOrder = array_reverse($stepOrder);
+        foreach ($reversedStepOrder as $lastTouchedStep) {
+            if ($jobApplicationSteps[$lastTouchedStep] === 'complete' || $jobApplicationSteps[$lastTouchedStep] === 'error') {
+                return redirect()->route('application.timeline', ['jobApplication' => $jobApplication, 'step' => $lastTouchedStep]);
+            }
+        }
+
+        // If no step has been reached yet it will take applicant to welcome step.
         return view('applicant/application-timeline-root')
```

Comment:
This won't necessarily take you to the welcome step - it'll take you to whichever step you wanted to go to. That means that you _can_ skip ahead to a future step, if you haven't clicked through any of the application yet.
This suggest this show() function contains enough logic that it should have its own unit tests.

Reply:
Hey, so I added in the tests. I had to make some new routes for the application timeline in the applicant group, since the demo routes don't work. After you check out the tests, we can just comment out everything and make a note to uncomment later. We can also just change the paths of the route to something different then the old applications route, I believe.